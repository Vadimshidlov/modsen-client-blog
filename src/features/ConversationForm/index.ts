export { ConversationForm } from '@/features/ConversationForm/ui/ConversationForm';
