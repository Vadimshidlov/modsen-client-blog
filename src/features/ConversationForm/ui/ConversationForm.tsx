'use client';

import React, { useEffect, useRef } from 'react';
import Image from 'next/image';
import emailjs from '@emailjs/browser';
import { toast } from 'react-toastify';
import styles from '@/features/ConversationForm/ui/ConversationForm.module.scss';
import { Inter } from 'next/font/google';
import { Button } from '@/shared/ui/Button';
import { CATEGORIES } from '@/shared/constants/categories';
import selectIconSrc from '@/features/ConversationForm/assets/select-icon.svg';
import { ErrorMessage, Field, Form, Formik, FormikHelpers, FormikValues } from 'formik';
import { conversationFormScheme } from '@/features/ConversationForm/constants/scheme';

const inter = Inter({
  subsets: ['latin'],
  weight: '400',
  variable: '--font-inter',
});

export type ConversationFormType = {
  userName: string;
  email: string;
  query: string;
  message: string;
};

export function ConversationForm() {
  const formRef = useRef<HTMLFormElement>(null);

  useEffect(() => emailjs.init(process.env.NEXT_PUBLIC_EMAILJS_PUBLIC_KEY as string), []);

  const sendForm = (
    data: FormikValues,
    { setSubmitting, resetForm }: FormikHelpers<ConversationFormType>,
  ) => {
    if (!formRef.current) return;

    emailjs
      .send(
        process.env.NEXT_PUBLIC_EMAILJS_SERVICE_ID as string,
        process.env.NEXT_PUBLIC_EMAILJS_TEMPLATE_ONE as string,
        {
          to_name: 'Vadim Shydlou',
          userName: data.userName,
          userEmail: data.email,
          query: data.query,
          message: data.message,
        },
      )
      .then(
        () => {
          toast.success('The data has been successfully sent!');
          resetForm();
        },
        () => {
          toast.error('Error during data sending!');
        },
      )
      .finally(() => {
        setSubmitting(false);
      });
  };

  return (
    <Formik
      validationSchema={conversationFormScheme}
      initialValues={{ userName: '', email: '', message: '', query: 'Startup' }}
      onSubmit={sendForm}
    >
      {({ isSubmitting }) => (
        <Form className={styles.conversationForm} ref={formRef}>
          <label htmlFor="userName">
            <Field
              className={`${styles.textInput} ${inter.variable}`}
              type="text"
              name="userName"
              placeholder="Full Name"
            />
            <ErrorMessage className={styles.validationError} name="userName" component="div" />
          </label>
          <label htmlFor="email">
            <Field
              className={`${styles.textInput} ${inter.variable}`}
              type="email"
              name="email"
              placeholder="Your Email"
            />
            <ErrorMessage className={styles.validationError} name="email" component="div" />
          </label>
          <div className={styles.selectCategoryWrapper}>
            <Field as="select" className={styles.selectCategory} name="query">
              {CATEGORIES.map(({ title }) => (
                <option className="select__item" value={title} key={title}>
                  {title}
                </option>
              ))}
            </Field>
            <Image className={styles.selectIcon} src={selectIconSrc} alt="select-icon" />
          </div>
          <label htmlFor="message">
            <Field
              as="textarea"
              className={`${styles.textInput} ${styles.conversationMessage}`}
              name="message"
              id="message"
              placeholder="Message"
            />
            <ErrorMessage className={styles.validationError} name="message" component="div" />
          </label>
          <Button isFormButton isSubmitting={isSubmitting} disabled={isSubmitting}>
            {isSubmitting ? 'Loading...' : 'Send Message'}
          </Button>
        </Form>
      )}
    </Formik>
  );
}
