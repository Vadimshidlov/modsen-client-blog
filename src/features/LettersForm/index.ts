export { LettersForm } from '@/features/LettersForm/ui/LettersForm';
