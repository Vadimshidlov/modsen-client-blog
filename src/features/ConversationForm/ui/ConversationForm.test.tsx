import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { ConversationForm } from '@/features/ConversationForm';

jest.mock('next-intl', () => ({
  useTranslations: jest.fn().mockReturnValue((key: string) => `Mocked ${key}`),
}));

describe('ConversationForm component', () => {
  test('should matches snapshot', () => {
    const { container } = render(<ConversationForm />);
    expect(container).toMatchSnapshot();
  });

  it('should display validation errors if form fields are empty', async () => {
    const { getByText, getByRole } = render(<ConversationForm />);

    const submitButton = getByRole('button');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(getByText('Name is required!')).toBeInTheDocument();
      expect(getByText('Email is required!')).toBeInTheDocument();
      expect(getByText('Message is required!')).toBeInTheDocument();
    });
  });

  it('should not render form errors with valid data', async () => {
    const { getByTestId, queryByText } = render(<ConversationForm />);

    const nameInput = getByTestId('userName-input');
    const emailInput = getByTestId('email-input');
    const messageInput = getByTestId('message-input');

    fireEvent.change(nameInput, { target: { value: 'John Doe' } });
    fireEvent.change(emailInput, { target: { value: 'john@example.com' } });
    fireEvent.change(messageInput, { target: { value: 'This is a test message' } });

    await waitFor(() => {
      expect(queryByText('Name is too short')).toBeNull();
      expect(queryByText('Invalid email')).toBeNull();
      expect(queryByText('Message is too short')).toBeNull();
    });
  });
});
