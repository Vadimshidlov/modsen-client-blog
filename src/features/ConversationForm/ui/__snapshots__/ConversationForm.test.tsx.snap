// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationForm component should matches snapshot 1`] = `
<div>
  <form
    action="#"
    class="conversationForm"
    data-testid="conversation-form"
  >
    <label
      for="userName"
    >
      <input
        class="textInput variable"
        data-testid="userName-input"
        id="userName"
        name="userName"
        placeholder="Mocked namePlaceholder"
        type="text"
        value=""
      />
    </label>
    <label
      for="email"
    >
      <input
        class="textInput variable"
        data-testid="email-input"
        id="email"
        name="email"
        placeholder="Mocked emailPlaceholder"
        type="email"
        value=""
      />
    </label>
    <div
      class="selectCategoryWrapper"
    >
      <select
        class="selectCategory"
        data-testid="conversation-select"
        name="query"
      >
        <option
          class="select__item"
          value="Business"
        >
          Business
        </option>
        <option
          class="select__item"
          value="Startup"
        >
          Startup
        </option>
        <option
          class="select__item"
          value="Economy"
        >
          Economy
        </option>
        <option
          class="select__item"
          value="Technology"
        >
          Technology
        </option>
      </select>
      <img
        alt="select-icon"
        class="selectIcon"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
    </div>
    <label
      for="message"
    >
      <textarea
        class="textInput conversationMessage"
        data-testid="message-input"
        id="message"
        name="message"
        placeholder="Mocked messagePlaceholder"
      />
    </label>
    <button
      class="button fullWidth"
      type="submit"
    >
      Mocked buttonText
    </button>
  </form>
</div>
`;
